### PR TITLE
Issue #14881: refactor summaryJavaDocCheck

### PR DESCRIPTION
**Pull request by mohitsatr**
resolves : #14881 

- remove null usage from `getFirstSentenceOrNull` & `findSentenceEndingOrNull`

### CODE REVIEW

**Comment by mohitsatr** [comment]
Github, generate report for SummaryJavadoc/all-examples-in-one

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java [hunk 1]
```diff
@@ -601,22 +609,23 @@ private static String getStringInsideTag(String result, DetailNode detailNode) {
      *
      * @param ast The Javadoc root node.
      * @param period The configured period symbol.
-     * @return The first sentence up to and excluding the period, or null if no ending was found.
+     * @return The first sentence up to and including the period,
+     * or empty string if no ending was found.
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java:613]: This is not good, as you force all users to read javadoc to understand how it works.
Please return Optional it will be very explicit on how method works.

**Review by romani — CHANGES_REQUESTED** [review]
Items

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java (hunk 1 above)
Comment (mohitsatr) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java:613]: done

Comment (mohitsatr) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java:613]: do we need to do the same with `findSentenceEnding` as well ?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java [hunk 2]
```diff
@@ -252,16 +257,20 @@ else if (containsForbiddenFragment(firstSentence)) {
     }
 
     /**
-     * Gets the node for the inline tag if present.
+     * Gets the index of the node for the inline tag if present.
      *
      * @param javadoc javadoc root node.
-     * @return the node for the inline tag if present.
+     * @return the index of node for the inline tag if present else -1.
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java:263]: revert this change. Keep optional.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java [hunk 3]
```diff
@@ -237,11 +242,11 @@ private void validateUntaggedSummary(DetailNode ast) {
...
         else if (!period.isEmpty()) {
             if (summaryDoc.contains(period)) {
-                final String firstSentence = getFirstSentenceOrNull(ast, period);
-                if (firstSentence == null) {
+                final Optional<String> firstSentence = getFirstSentence(ast, period);
+                if (firstSentence.isEmpty()) {
                     log(ast.getLineNumber(), MSG_SUMMARY_FIRST_SENTENCE);
                 }
-                else if (containsForbiddenFragment(firstSentence)) {
+                else if (containsForbiddenFragment(firstSentence.orElseThrow())) {
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java:249]: why orElseThrow ?

Check should not throw exceptions, except for cases of wrong config and bad AST.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java [hunk 4]
```diff
@@ -646,12 +658,12 @@ private static Stream<String> streamTextParts(DetailNode node) {
      *
      * @param text The string to search.
      * @param period The period character to find.
-     * @return The string up to and excluding the period, or null if no ending was found.
+     * @return The string up to and including the period,
+     *     or empty string if no ending was found.
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java:662]: please return Optioonal.

**Review by romani — CHANGES_REQUESTED** [review]
items:

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java (hunk 1 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java]: yes, please.

**Comment by mohitsatr** [comment]
hi @romani 
 
Screenshot from 2024-11-30 10-20-05

build was working fine before I rebased the branch with master. could you confirm if it's only on my side ?

**Comment by romani** [comment]
Try to git reset to few commits back in history, and do build.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java (hunk 3 above)
Comment (mohitsatr) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java]: fixed

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java (hunk 4 above)
Comment (mohitsatr) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java (hunk 2 above)
Comment (mohitsatr) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java [hunk 5]
```diff
@@ -210,19 +208,25 @@ public int[] getRequiredJavadocTokens() {
...
-            && isSummaryTag(inlineTagNode)
-            && isDefinedFirst(inlineTagNode)) {
-            validateSummaryTag(inlineTagNode);
-        }
-        else if (inlineTag.isPresent() && isInlineReturnTag(inlineTagNode)) {
-            validateInlineReturnTag(inlineTagNode);
-        }
-        else if (!startsWithInheritDoc(ast)) {
+        final Optional<DetailNode> inlineTagNode = getInlineTagNode(ast);
+        final AtomicBoolean shouldValidateUntaggedSummary = new AtomicBoolean(true);
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java:212]: just boolean

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java [hunk 6]
```diff
@@ -210,19 +208,25 @@ public int[] getRequiredJavadocTokens() {
...
+
+        inlineTagNode.ifPresent(node -> {
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java:214]: can we do 
```
.map(some function)
.orElse(some function);
```

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java [hunk 7]
```diff
@@ -237,8 +241,9 @@ private void validateUntaggedSummary(DetailNode ast) {
         }
         else if (!period.isEmpty()) {
             if (summaryDoc.contains(period)) {
-                final String firstSentence = getFirstSentenceOrNull(ast, period);
-                if (firstSentence == null) {
+                final String firstSentence = getFirstSentence(ast, period).orElse("");
+
+                if (firstSentence.isEmpty()) {
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java:246]: can we skip `.orElse("")` ?
to use `if (firstSentence.ifPresent()) {`

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java [hunk 8]
```diff
@@ -255,13 +260,17 @@ else if (containsForbiddenFragment(firstSentence)) {
...
-        return Arrays.stream(javadoc.getChildren())
-            .filter(SummaryJavadocCheck::isInlineTagPresent)
-            .findFirst()
-            .map(SummaryJavadocCheck::getInlineTagNodeForAst);
+        Optional<DetailNode> result = Optional.empty();
+        for (DetailNode child : javadoc.getChildren()) {
+            if (isInlineTagPresent(child)) {
+                result = Optional.of(getInlineTagNodeForAst(child));
+                break;
+            }
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java:271]: why we  need this refactoring ?

**Review by romani — CHANGES_REQUESTED** [review]
items:

**Comment by mohitsatr** [comment]
null usage in `getInlineTagNodeForAst()` as well

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java (hunk 7 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java:246]: > ifPresentOrElse api

Sure

**Comment by romani** [comment]
>null usage in getInlineTagNodeForAst() as well

Can do this in separate PR?
This is already very complicated update, let's do step by step.

**Comment by mohitsatr** [comment]
> Can do this in separate PR?
sure, i'll send another pr for that.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java (hunk 6 above)
Comment (mohitsatr) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java:214]: tried something like : 

        final boolean shouldValidateUntaggedSummary = inlineTagNode
                .map(this::validateInlineTagNode)
                .orElse(false);

        if (shouldValidateUntaggedSummary && !startsWithInheritDoc(ast)) {
            validateUntaggedSummary(ast);
        }
        
and got `NAB_NEEDLESS_BOOLEAN_CONSTANT_CONVERSION` for `.orElse(false);`

Comment (mohitsatr) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java:214]: how about this approach : 
```
public void visitJavadocToken(DetailNode ast) {
    final Optional<DetailNode> inlineTagNode = getInlineTagNode(ast);
    AtomicBoolean shouldValidateUntaggedSummary = new AtomicBoolean(true);

    inlineTagNode.ifPresent(node -> {
         shouldValidateUntaggedSummary.set(validateInlineTagNode(node));
    });
      if (shouldValidateUntaggedSummary.get() && !startsWithInheritDoc(ast)) {
            validateUntaggedSummary(ast);
        }

    }
```
can't do without AtomicBoolean .

**Comment by romani** [comment]
Please keep pulling code  from main upstream repository master branch, to let CI be green.

**Comment by romani** [comment]
This PR becomes too complicated, if you see any benefits to split each method refactoring in separate PR, please consider it. Reviewing is not easier on smaller update, so merging if quicker.

**Comment by mohitsatr** [comment]
> This PR becomes too complicated, if you see any benefits to split each method refactoring in separate PR, please consider it. Reviewing is not easier on smaller update, so merging if quicker.

alright. im closing this pr in favour of raising separate pr for each method.
